### PR TITLE
Fix #1243: `read` raises EOFError on false inputs

### LINKED ISSUE #1243: read-str raises EOFError given "\"\"", "b\"\"", "[]", or "{}"
Or `"#{}"`.

### PR DESCRIPTION
**Pull request by Kodiologist**

### CODE REVIEW

**Comment by Kodiologist** [comment]
I found a very similar bug in `eval`, so I added a fix for that. Remember not to squash.

File: hy/compiler.py [hunk 1]
```diff
@@ -2689,18 +2689,20 @@ def hy_compile(tree, module_name, root=ast.Module, get_expr=False):
     body = []
     expr = None
 
-    if tree:
+    if not (isinstance(tree, HyObject) or type(tree) is list):
```
Comment (kirbyfan64) [inline hy/compiler.py]: Is this really that useful? IMO part of the charm of Python is duck-typing, where someone can pass a custom but API-compatible list subclass, and it "just works". Can't you just do `isinstance(tree, list)`?

Comment (Kodiologist) [inline hy/compiler.py]: If you passed in an object that is neither exactly `list` nor a `HyObject`, the below conditional in the assignment to `spoof_tree` would do the wrong thing.

Comment (Kodiologist) [inline hy/compiler.py]: Also, this is an internal function. If it receives an argument that's of a type we didn't expect, that's a bug.

Comment (kirbyfan64) [inline hy/compiler.py]: Fair enough.

**Comment by Kodiologist** [comment]
Thanks. I'm not going to ask you to undo the merge or reopen the PR, but I think this wasn't a trivial change, so it should've gotten a second review.

**Comment by kirbyfan64** [comment]
Dammit, that was an accident...

GitHub mobile is killing me here...

**Pull request merged**